sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala

```diff
@@ -192,6 +192,9 @@ case class ExpressionEncoder[T](
 
     if (isSerializedAsStructForTopLevel) {
       val nullSafeSerializer = objSerializer.transformUp {
+        case _ @ AssertNotNull(a: AssertNotNull, _) =>
+          // strip redundant AssertNotNull
+          a
         case r: BoundReference =>
```

Comment:
Actually, I'm thinking about whether we can remove this directly. It seems we always do `AssertNotNull` for Product when we construct serializer:

```
val fieldValue = Invoke(
    AssertNotNull(inputObject, walkedTypePath), fieldName, dataTypeFor(fieldType),
    returnNullable = !fieldType.typeSymbol.asClass.isPrimitive)
```

Reply:
I think we should strip `AssertNotNull` there:
```
val fieldValue = Invoke(
    inputObject, fieldName, dataTypeFor(fieldType),
    returnNullable = !fieldType.typeSymbol.asClass.isPrimitive)
```
The `fieldValue` is used in `nonNullOutput` in
```
expressions.If(IsNull(inputObject), nullOutput, nonNullOutput)
```

So `inputObject` can't be null already

Reply:
Hi, @cloud-fan . Thanks for your suggestion. Yeah, this actually could be an equivalent solution.  But, we may lost `walkedTypePath`(though, current fix lost it, either) info, would it be OK ?



 > `expressions.If(IsNull(inputObject), nullOutput, nonNullOutput)`
So inputObject can't be null already

Not sure I got your point clearly. Just to be clarify, we'll strip `If` expression later after we added `AssertNotNull` expression:

```
nullSafeSerializer match {
  case If(_: IsNull, _, s: CreateNamedStruct) => s
  case _ =>
     throw new RuntimeException(s"class $clsName has unexpected serializer: $objSerializer")
}
```
So, it seems not appropriate to say `already` at this point(wihout `AssertNotNull`), or maybe I misunderstand your opinion.

Reply:
well, if you look at it globally, yes, we may strip the `If` later and `inputObject` can be null.

However, I think we should make the code less coupled with each other. When we create the `If`, we can safely make the assumption that `inputObject` is not null, as at that point we don't know it will be stripped. When we strip the `If`, we should make sure the assumption still holds, by inserting new `AssertNotNull`s.

Reply:
Thanks for your explanation, quite clearly and good!! And how about the `walkedTypePath` I mentioned above ? If we strip `AssertNotNull` there we may lost it,  would it be OK ?

Reply:
it's ok and it will be replaced by `top level Product or row object`, which is more corrected to me.

Reply:
Okay, I'll update soon.